File: sdk/go/internal/providerworkflow/input.go
package providerworkflow

import (
	"bytes"
	"encoding/json"
	"errors"
	"io"
	"net/url"
	"os"
	"path/filepath"
	"runtime"
	"strconv"
)

type Target struct {
	Service         string `json:"service"`
	Route           string `json:"route"`
	Contract        string `json:"contract"`
	Function        string `json:"function"`
	Publication     string `json:"publication"`
	ComponentDigest string `json:"componentDigest"`
}

type Input struct {
	SchemaVersion    string            `json:"schemaVersion"`
	Language         string            `json:"language"`
	Endpoint         string            `json:"endpoint"`
	Tenant           string            `json:"tenant"`
	CredentialFile   string            `json:"credentialFile"`
	ControlDirectory string            `json:"controlDirectory"`
	UpstreamURL      string            `json:"upstreamUrl"`
	PolicyDocument   string            `json:"policyDocument"`
	Targets          map[string]Target `json:"targets"`
}

func Load(path string) (Input, string, error) {
	var input Input
	if runtime.GOOS != "linux" || !filepath.IsAbs(path) {
		return input, "", errors.New("participant-requires-linux-and-absolute-config")
	}
	data, failure := readProtected(path, 32768, false)
	if failure != nil || decodeJSON(data, &input) != nil {
		return input, "", errors.New("participant-invalid-protected-input")
	}
	if input.SchemaVersion != "latent.sdk.provider.workflow.input.v1" || input.Language != "go" || input.Tenant != "tests" ||
		len(input.Targets) != 3 || len(input.PolicyDocument) > 16384 || !filepath.IsAbs(input.ControlDirectory) {
		return input, "", errors.New("participant-invalid-input-profile")
	}
	for _, name := range []string{"http", "blob", "callee"} {
		target, present := input.Targets[name]
		if !present || target.Service == "" || target.Route == "" || target.Contract == "" || target.Function == "" || target.Publication == "" || target.ComponentDigest == "" {
			return input, "", errors.New("participant-missing-exact-target")
		}
	}
	origin, failure := url.Parse(input.UpstreamURL)
	if failure != nil {
		return input, "", errors.New("participant-invalid-fixture-upstream")
	}
	port, failure := strconv.ParseUint(origin.Port(), 10, 16)
	if failure != nil || port == 0 || input.UpstreamURL != "http://localhost:"+strconv.FormatUint(port, 10)+"/allowed" {
		return input, "", errors.New("participant-invalid-fixture-upstream")
	}
	control, failure := os.Lstat(input.ControlDirectory)
	if failure != nil || !control.IsDir() || control.Mode().Perm() != 0700 {
		return input, "", errors.New("participant-control-directory-not-private")
	}
	var policy struct {
		FormatVersion int               `json:"formatVersion"`
		Tenant        string            `json:"tenant"`
		Rules         []json.RawMessage `json:"rules"`
	}
	if decodeJSON([]byte(input.PolicyDocument), &policy) != nil || policy.FormatVersion != 1 || policy.Tenant != input.Tenant || policy.Rules == nil || len(policy.Rules) != 0 {
		return input, "", errors.New("participant-policy-must-grant-no-authority")
	}
	token, failure := readProtected(input.CredentialFile, 512, true)
	if failure != nil || len(token) == 0 {
		return input, "", errors.New("participant-invalid-protected-credential")
	}
	return input, string(token), nil
}

func readProtected(path string, maximum int64, credential bool) ([]byte, error) {
	if !filepath.IsAbs(path) {
		return nil, errors.New("absolute-input-path-required")
	}
	parent, failure := os.Lstat(filepath.Dir(path))
	if failure != nil || !parent.IsDir() || parent.Mode().Perm() != 0700 {
		return nil, errors.New("private-input-directory-required")
	}
	before, failure := os.Lstat(path)
	if failure != nil || !before.Mode().IsRegular() || before.Size() > maximum || (credential && before.Mode().Perm() != 0600) {
		return nil, errors.New("invalid-protected-input-file")
	}
	file, failure := os.Open(path)
	if failure != nil {
		return nil, errors.New("protected-input-open-failed")
	}
	defer file.Close()
	opened, failure := file.Stat()
	if failure != nil || !os.SameFile(before, opened) || opened.Mode() != before.Mode() {
		return nil, errors.New("protected-input-changed")
	}
	data, failure := io.ReadAll(io.LimitReader(file, maximum+1))
	if failure != nil || int64(len(data)) > maximum {
		return nil, errors.New("protected-input-exceeds-bound")
	}
	return data, nil
}

func decodeJSON(data []byte, target any) error {
	structure := json.NewDecoder(bytes.NewReader(data))
	structure.UseNumber()
	remaining := 2048
	if uniqueJSON(structure, 0, &remaining) != nil {
		return errors.New("invalid-closed-json-input")
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	if failure := decoder.Decode(target); failure != nil {
		return errors.New("invalid-closed-json-input")
	}
	var extra any
	if decoder.Decode(&extra) != io.EOF {
		return errors.New("trailing-json-input")
	}
	return nil
}

func uniqueJSON(decoder *json.Decoder, depth int, remaining *int) error {
	*remaining--
	if depth > 16 || *remaining < 0 {
		return errors.New("json-structure-exceeds-bound")
	}
	value, failure := decoder.Token()
	if failure != nil {
		return failure
	}
	delimiter, container := value.(json.Delim)
	if !container {
		return nil
	}
	if delimiter != '{' && delimiter != '[' {
		return errors.New("invalid-json-container")
	}
	keys := make(map[string]bool)
	for decoder.More() {
		if delimiter == '{' {
			key, failure := decoder.Token()
			if failure != nil {
				return failure
			}
			name, valid := key.(string)
			if !valid || keys[name] {
				return errors.New("duplicate-json-field")
			}
			keys[name] = true
		}
		if failure := uniqueJSON(decoder, depth+1, remaining); failure != nil {
			return failure
		}
	}
	_, failure = decoder.Token()
	return failure
}

File: sdk/dotnet/Latent.Sdk.ProviderWorkflow/Input.cs
using System.Net;
using System.Text;
using System.Text.Json;
using Latent.Sdk.Transport;
using Profile = Latent.Sdk.Profile;

namespace Latent.Examples;

internal sealed record Target(string Service, string Route, string Contract, string Function, string Publication, string ComponentDigest);

internal sealed class Input
{
    internal const string MediaType = "application/vnd.latent.wit-values.v1+json";
    internal string Endpoint { get; }
    internal string Tenant { get; }
    internal string Control { get; }
    internal string Upstream { get; }
    internal string PolicyDocument { get; }
    internal IReadOnlyDictionary<string, Target> Targets { get; }
    private readonly string bearer;

    internal Input(string path)
    {
        Require(OperatingSystem.IsLinux() && Path.IsPathFullyQualified(path), "protected-linux-input-required");
        Private(path, false);
        string directory = Path.GetDirectoryName(path)!;
        Private(directory, true);
        using JsonDocument document = JsonDocument.Parse(Read(path, 65536), new JsonDocumentOptions { MaxDepth = 10 });
        JsonElement root = document.RootElement;
        Shape(root, "schemaVersion", "language", "endpoint", "tenant", "credentialFile", "controlDirectory", "upstreamUrl", "targets", "policyDocument");
        Require(Field(root, "schemaVersion") == "latent.sdk.provider.workflow.input.v1" && Field(root, "language") == "dotnet", "input-profile");
        Endpoint = Field(root, "endpoint");
        Tenant = Field(root, "tenant");
        Require(Uri.TryCreate(Endpoint, UriKind.Absolute, out Uri? endpoint) && endpoint.Scheme == "http" &&
            IPAddress.TryParse(endpoint.DnsSafeHost, out IPAddress? address) && IPAddress.IsLoopback(address), "numeric-loopback-endpoint");
        string credential = Field(root, "credentialFile", 4096);
        Control = Field(root, "controlDirectory", 4096);
        Require(Path.IsPathFullyQualified(credential) && Path.GetDirectoryName(credential) == directory &&
            Path.GetFullPath(Control) == Path.Combine(Path.GetDirectoryName(directory)!, "control"), "protected-input-paths");
        Private(credential, false);
        Private(Control, true);
        byte[] token = Read(credential, 512);
        Require(token.Length != 0 && token.All(value => value is >= 33 and <= 126), "credential-shape");
        bearer = Encoding.ASCII.GetString(token);
        Upstream = Field(root, "upstreamUrl");
        Require(Uri.TryCreate(Upstream, UriKind.Absolute, out Uri? upstream) && upstream.Scheme == "http" &&
            upstream.Host == "localhost" && upstream.Port > 0 && Upstream == $"http://localhost:{upstream.Port}/allowed", "provider-target");
        PolicyDocument = Field(root, "policyDocument", 16384);
        using JsonDocument policy = JsonDocument.Parse(PolicyDocument, new JsonDocumentOptions { MaxDepth = 4 });
        Shape(policy.RootElement, "formatVersion", "tenant", "rules");
        Require(policy.RootElement.GetProperty("formatVersion").GetInt32() == 1 && Field(policy.RootElement, "tenant") == Tenant &&
            policy.RootElement.GetProperty("rules").ValueKind == JsonValueKind.Array && policy.RootElement.GetProperty("rules").GetArrayLength() == 0,
            "non-granting-policy-required");
        JsonElement targets = root.GetProperty("targets");
        Shape(targets, "http", "blob", "callee");
        var parsed = new Dictionary<string, Target>();
        foreach (string name in new[] { "http", "blob", "callee" })
        {
            JsonElement selected = targets.GetProperty(name);
            Shape(selected, "service", "route", "contract", "function", "publication", "componentDigest");
            parsed.Add(name, new(Field(selected, "service"), Field(selected, "route"), Field(selected, "contract"),
                Field(selected, "function"), Field(selected, "publication"), Field(selected, "componentDigest")));
        }
        Targets = parsed;
    }

    public override string ToString() => "Latent provider workflow input (redacted)";

    internal ClientOptions Options(bool wrongCredential = false, bool limited = false) => new()
    {
        Endpoint = Endpoint, BearerToken = wrongCredential ? "LSF-NOT-A-VALID-CREDENTIAL" : bearer,
        DefaultTimeout = TimeSpan.FromSeconds(3), ConnectTimeout = TimeSpan.FromSeconds(2),
        MaxInFlight = 4, ReservedRecovery = 2, MaxQueued = 2, MaxResponseBytes = limited ? 8 : 1024 * 1024
    };

    internal Profile.InvokeRequest Request(string provider, string identity, string? function = null, string? tenant = null, ulong? deadline = null)
    {
        Target target = Targets[provider];
        bool callee = provider == "callee";
        ReadOnlyMemory<byte> payload = callee ? "[]"u8.ToArray() :
            JsonSerializer.SerializeToUtf8Bytes(new object[] { 0, provider == "http" ? Upstream : "", "0" });
        return new(identity, null, null, new(tenant ?? Tenant, target.Service, target.Contract, function ?? target.Function, target.Route),
            payload, MediaType, deadline, 0, null,
            new(function == "spin" ? 1000UL : callee ? 100_000_000UL : 10_000_000_000UL, callee ? 4194304UL : 16777216UL,
                0, callee ? 0U : 8U, 0, 0, provider == "blob" ? 65536UL : 0UL, provider == "blob" ? 65536UL : 0UL, 0, 0, 5000),
            new Dictionary<string, string>());
    }

    internal void Receipt(Profile.InvokeResponse response, string provider, string identity)
    {
        Target target = Targets[provider];
        Require(response.ActivationId == identity && response.PublicationId == target.Publication && response.ReleaseDigest == target.ComponentDigest &&
            response.RevisionId.Length != 0 && response.RouteGeneration != 0 && response.Consumption is not null, "invocation-receipt");
    }

    internal static ulong Guest(Profile.InvokeResponse response)
    {
        Require(response.Success is not null && response.Success.MediaType == MediaType, "typed-guest-success");
        using JsonDocument value = JsonDocument.Parse(response.Success!.Payload, new JsonDocumentOptions { MaxDepth = 4 });
        Require(value.RootElement.ValueKind == JsonValueKind.Array && value.RootElement.GetArrayLength() == 1 &&
            value.RootElement[0].ValueKind == JsonValueKind.String, "canonical-unsigned-guest-result");
        return Profile.UnsignedDecimal.Parse(value.RootElement[0].GetString()!);
    }

    internal static byte[] Read(string path, int maximum)
    {
        using var source = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read);
        Require(source.Length <= maximum, "input-byte-bound");
        byte[] data = new byte[checked((int)source.Length)];
        source.ReadExactly(data);
        Require(source.ReadByte() == -1, "input-byte-bound");
        return data;
    }

    internal static void Private(string path, bool directory)
    {
        FileAttributes attributes = File.GetAttributes(path);
        Require((attributes & FileAttributes.ReparsePoint) == 0 && ((attributes & FileAttributes.Directory) != 0) == directory,
            "private-input-kind");
        if (!OperatingSystem.IsLinux()) throw new InvalidOperationException("protected-linux-input-required");
        UnixFileMode mode = File.GetUnixFileMode(path);
        const UnixFileMode shared = UnixFileMode.GroupRead | UnixFileMode.GroupWrite | UnixFileMode.GroupExecute |
            UnixFileMode.OtherRead | UnixFileMode.OtherWrite | UnixFileMode.OtherExecute;
        Require((mode & shared) == 0 && (mode & UnixFileMode.UserRead) != 0, "private-input-permissions");
    }

    private static void Shape(JsonElement value, params string[] fields)
    {
        Require(value.ValueKind == JsonValueKind.Object, "input-object");
        string[] actual = value.EnumerateObject().Select(property => property.Name).ToArray();
        Require(actual.Length == fields.Length && actual.Distinct().Count() == actual.Length && fields.All(actual.Contains), "input-object-fields");
    }

    private static string Field(JsonElement value, string name, int maximum = 256)
    {
        JsonElement field = value.GetProperty(name);
        Require(field.ValueKind == JsonValueKind.String, "input-string");
        string text = field.GetString()!;
        Require(text.Length is > 0 && text.Length <= maximum && !text.Any(character => char.IsControl(character)), "input-string-bound");
        return text;
    }

    internal static void Require(bool condition, string reason)
    {
        if (!condition) throw new InvalidOperationException(reason);
    }
}
